### PR TITLE
6.0.0 Beta branch

### PR DESCRIPTION
**Pull request by bcomnes**
A branch to track beta releases for 6.0.0

```
$ npm i nanomorph@beta --save
```

### CODE REVIEW

**Comment by kristoferjoseph** [comment]
This patch fixes one case where a `proxy-node` was being inserted when `nanomorph` could not find a matching node.
https://github.com/choojs/nanomorph/blob/beta/index.js#L140

There must be another case where this is happening though since I am still able to find `proxy-nodes` inserted into children of list elements.

**Comment by kristoferjoseph** [comment]
OK so what I've learned is that there are only very few cases where you want to insert the real node from the proxy node. 

There are two places where we need an extra case.
1. when appending a node when there is no oldChild and your newChild is a proxy
   a. doing this every time can cause a worst case endless loop
2. when inserting at the index when we can't find a match
  b. doing this every time can insert random nodes in the wrong place

**Comment by kristoferjoseph** [comment]
Issue remains that inserting the realNode from a proxy can enter into wrong places in the DOM.
You can end up with DOM exceptions like new child element contains the parent.

@bcomnes would be cool to brainstorm more on how to avoid inserting realNodes from proxies into the wrong index.

File: index.js [hunk 1]
```diff
@@ -94,7 +98,11 @@ function updateChildren (newNode, oldNode) {
 
     // There is no old child, add new
     } else if (!oldChild) {
-      oldNode.appendChild(newChild)
+      if (isProxy(newChild) && newChild.realNode) {
+        break
```
Comment (kristoferjoseph) [inline index.js]: without this you can end up crashing the browser

File: index.js [hunk 2]
```diff
@@ -133,7 +141,11 @@ function updateChildren (newNode, oldNode) {
 
       // Insert the node at the index if we couldn't morph or find a matching node
       } else {
-        oldNode.insertBefore(newChild, oldChild)
+        if (isProxy(newChild) && !newChild.isSameNode(oldChild) && newChild.realNode) {
+          oldNode.insertBefore(newChild.realNode, oldChild)
```
Comment (kristoferjoseph) [inline index.js]: We need some kind of check in here to make sure we aren't inserting a node in the wrong place.

Comment (yoshuawuyts) [reply index.js]: we could cheat and add a magic check to make sure we never insert proxy nodes perhaps?

Comment (yoshuawuyts) [reply index.js]: might create other fun (read: critical) failure scenarios tho, haha

Comment (kristoferjoseph) [reply index.js]: @yoshuawuyts yes. I had a version where it gave up every time it was about to insert a proxy, but it left things out of the DOM :/

so there are times when you do want to insert that realNode, just hard to know what the heuristics should be.

**Comment by kristoferjoseph** [comment]
Was able to get this all working with a the myers diffing algorithm written by @graforlock, but ran into an issue that seems like a deal breaker... when morphing a parent node the proxy nodes for the children are being copied and appended. :/

Ever returning a proxy node doesn't appear to be a viable option based on these findings. Seems like we need a different approach of a two pass system ( which @graforlock's impl has affordances for ) where we check for nodes that need updates against a proxy tree then apply the changes to the real tree, never inserting a proxy node.

basically making a copy dom... or proxy tree... virtual dom :/ to get around the unnecessary removal of child elements.

**Comment by bcomnes** [comment]
Hehe fun!  Thanks for digging into this.

**Comment by bcomnes** [comment]
I'm convinced this was a dead end.  I'll be decommissioning this and marking the tag depreciated.
